### PR TITLE
Replace ECDSA and Schnorr with Ed25519

### LINKED ISSUE #146: Remove SchnorrSignature
Remove `SchnorrSignature` from `key` package and replace it with `Ed25519Signature`.

### PR DESCRIPTION
**Pull request by HoOngEe**
It resolves #146

### CODE REVIEW

**Comment by HoOngEe** [comment]
In this pr the e2e test cannot fully covered because this change requires SDK changes.

File: key/src/ed25519_private.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::ops::Deref;
+use std::str::FromStr;
+
```
Comment (sgkim126) [inline key/src/ed25519_private.rs]: Please remove the blank lines.
https://github.com/CodeChain-io/codechain/issues/1940

File: key/src/ed25519_private.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    }
+}
+
+impl From<SecretKey> for Private {
+    fn from(k: SecretKey) -> Self {
+        Private(k)
+    }
+}
+
+impl Encodable for Private {
```
Comment (sgkim126) [inline key/src/ed25519_private.rs]: Please add a test for rlp encoding.

File: key/src/ed25519_public.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+impl From<PublicKey> for Public {
+    fn from(k: PublicKey) -> Self {
+        Public(k)
+    }
+}
+
+impl Encodable for Public {
+    fn rlp_append(&self, s: &mut RlpStream) {
+        s.append_single_value(&self.as_ref());
```
Comment (sgkim126) [inline key/src/ed25519_public.rs]: Why does it use `append_single_value`, while you used `append` for `Private`?

File: core/src/account_provider.rs [hunk 4]
```diff
@@ -264,7 +264,7 @@ mod tests {
     fn unlock_account_temp() {
         let kp = Random.generate().unwrap();
         let ap = AccountProvider::transient_provider();
-        assert!(ap.insert_account(*kp.private(), &"test".into()).is_ok());
```
Comment (sgkim126) [inline core/src/account_provider.rs]: Why do you prohibit copy?

File: key/src/lib.rs [hunk 5]
```diff
@@ -26,52 +26,41 @@ extern crate rand_xorshift;
...
 extern crate serde;
 #[macro_use]
 extern crate serde_derive;
 extern crate serde_json;
+extern crate sodiumoxide;
 
 mod address;
-mod ecdsa;
+mod ed25519;
+mod ed25519_private;
```
Comment (sgkim126) [inline key/src/lib.rs]: How about move `ed25519_private` and`ed25519_public` under the `ed25519` module?

File: key/src/ed25519_private.rs (hunk 1 above)
Comment (HoOngEe) [reply key/src/ed25519_private.rs]: Thank you, I'll keep it in mind.

File: key/src/ed25519_public.rs (hunk 3 above)
Comment (HoOngEe) [reply key/src/ed25519_public.rs]: Sorry, I should have used append_single_value for `Private`.
When I used `append` for the slice, I observed that the encoded `Public` is not separated as an independent list element in the bigger encoded structs.

Comment (HoOngEe) [reply key/src/ed25519_public.rs]: Even, I don't have to implement `Encodable` and `Decodable` for `Private`. I'll modify it.

File: core/src/account_provider.rs (hunk 4 above)
Comment (HoOngEe) [reply core/src/account_provider.rs]: It is a very low-level security concern and `Rust` can safely dispose private information by implementing memory clear logic in `Drop` trait. However, the more conservative way is prohibiting copy behavior of private information. Previously, private information was stored in `H256` type and the `DecryptedAccount` which possibly contains private key info implements memory-clearing `Drop` trait.
However, I could not sure about other usages are desirable because `H256` was used just for hash information also. And this change did not require many changes. What do you think of it?

File: key/src/lib.rs (hunk 5 above)
Comment (HoOngEe) [reply key/src/lib.rs]: I made `ed25519` module and put them into it.

File: core/src/account_provider.rs (hunk 4 above)
Comment (majecty) [reply core/src/account_provider.rs]: @HoOngEe How about adding summary of this description in the commit message?

File: core/src/account_provider.rs [hunk 6]
```diff
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
-use ckey::{public_to_address, Address, Error as KeyError, Generator, KeyPair, Password, Private, Public, Random};
+use ckey::{public_to_address, Address, Error as KeyError, Generator, Password, Private, Public, Random};
```
Comment (majecty) [inline core/src/account_provider.rs]: Great!
I'm happy to see this commit.

File: key/src/exchange.rs [hunk 7]
```diff
@@ -15,23 +15,34 @@
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
 use crate::{Error, Private, Public, SharedSecret};
-use sodiumoxide::crypto::kx::server_session_keys;
+use sodiumoxide::crypto::kx::{client_session_keys, server_session_keys};
 use std::convert::TryInto;
 
-pub fn exchange(other_public: &Public, my_private: &Private) -> Result<SharedSecret, Error> {
-    let my_public = &my_private.public_key();
+pub fn server_session_key(client_public: &Public, server_private: &Private) -> Result<SharedSecret, Error> {
```
Comment (majecty) [inline key/src/exchange.rs]: @HoOngEe I don't know what is the server_session_key and client_session_key, is it related to the Diffie-Hellman key exchange? Could you give me some material?

**Comment by HoOngEe** [comment]
The key exchange function supported by `sodiumoxide` operates like follows:
It calculate a key by ECDH: q = (local_private_key) * (remote_public_key). In `client_session_key` 
, local means the client side and remote means the server side and vice versa. Then it hashes (q || client_public_key || server_public_key) into 64 bytes. It splits the result as tuple. The `server_session_key` returns (hash[0:32], hash[33:64]) and `client_session_key` returns (hash[33:64], hash[0:32]). If we want a single shared secret, we have to use the Fst(tuple_resulted_from_sever_session_key) == Snd(tuple_resulted_from_client_session_key) or the other.

File: key/src/exchange.rs (hunk 7 above)
Comment (HoOngEe) [reply key/src/exchange.rs]: https://github.com/CodeChain-io/foundry/pull/154#issuecomment-586122289

File: core/src/consensus/stake/actions.rs [hunk 8]
```diff
@@ -127,20 +143,24 @@ impl Action {
...
                 let action = Action::ChangeParams {
                     metadata_seq: *metadata_seq,
                     params: params.clone(),
-                    signatures: vec![],
+                    approvals: vec![],
                 };
                 let encoded_action = H256::blake(rlp::encode(&action));
-                for signature in signatures {
+                for approval in approvals {
                     // XXX: Signature recovery is an expensive job. Should we do it twice?
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/stake/actions.rs]: Thank you I removed the comment

File: key/src/ed25519_private.rs (hunk 2 above)
Comment (HoOngEe) [reply key/src/ed25519_private.rs]: The `Encodable` for `Private` is removed

File: core/src/account_provider.rs (hunk 4 above)
Comment (sgkim126) [reply core/src/account_provider.rs]: I'm not sure the purpose yet. Prohibiting a copy of private data is not a way to keep data privately. I can agree that H256 seems to general and you want to replace it with `SecretKey`, but it doesn't mean that SecretKey is safer because it doesn't provide `Copy`.

File: key/src/lib.rs [hunk 9]
```diff
@@ -31,6 +31,7 @@ extern crate serde;
 #[macro_use]
 extern crate serde_derive;
 extern crate serde_json;
+extern crate sodiumoxide;
```
Comment (sgkim126) [inline key/src/lib.rs]: You don't need to add extern crate statements anymore since 2018 edition.

Comment (HoOngEe) [reply key/src/lib.rs]: Thank you, I studied the comment on it after you pointed out.

File: key/src/ed25519/ed25519_private.rs [hunk 10]
```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 Kodebox, Inc.
```
Comment (sgkim126) [inline key/src/ed25519/ed25519_private.rs]: I think this can be renamed to private.rs since it's under the ed25519 module.

File: key/src/ed25519/ed25519_public.rs [hunk 11]
```diff
@@ -0,0 +1,134 @@
+// Copyright 2020 Kodebox, Inc.
```
Comment (sgkim126) [inline key/src/ed25519/ed25519_public.rs]: Same as the above.

File: key/src/ed25519/ed25519_signature.rs [hunk 12]
```diff
@@ -0,0 +1,121 @@
+// Copyright 2018-2020 Kodebox, Inc.
```
Comment (sgkim126) [inline key/src/ed25519/ed25519_signature.rs]: ditto.

File: core/src/account_provider.rs (hunk 4 above)
Comment (HoOngEe) [reply core/src/account_provider.rs]: Oh my god. I misunderstood the types. The actual situation is: `SecretKey` by `sodiumoxide` implemented memory clearing `Drop` trait inside so the `Copy` marker is incompatible. Therefore, the `Private` which is a wrapper for `SecretKey` cannot mark `Copy` on it. In the sense, this change is inevitable. Is there some use cases need to allow `Copy` behavior?

File: key/src/ed25519/ed25519_private.rs (hunk 10 above)
Comment (HoOngEe) [reply key/src/ed25519/ed25519_private.rs]: Ok, that's more reasonable. I changed it.

File: core/src/account_provider.rs (hunk 4 above)
Comment (sgkim126) [reply core/src/account_provider.rs]: No, we don't need `Copy`, we can use `Clone` instead of it.

Comment (sgkim126) [reply core/src/account_provider.rs]: Please change the commit message.

File: key/src/ed25519/private.rs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::Public;
+use primitives::H512;
+use sodiumoxide::crypto::sign::{gen_keypair, SecretKey, SECRETKEYBYTES};
+use std::ops::Deref;
+use std::str::FromStr;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Private(SecretKey);
```
Comment (sgkim126) [inline key/src/ed25519/private.rs]: I think it's better to add a comment that `SecretKey` clears memory when it is dropped to inform others who want to replace the underlying type.

File: key/src/ed25519/private.rs [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+impl FromStr for Private {
+    type Err = crate::Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let for_slice = H512::from_str(s).map_err(|e| crate::Error::Custom(format!("{:?}", e)))?;
+        Ok(SecretKey::from_slice(&for_slice).expect("H512 has length 64").into())
+    }
+}
+
+impl Default for Private {
```
Comment (sgkim126) [inline key/src/ed25519/private.rs]: The only use case of `Default` of `Private` is `Drop` of `DecrptedAccount`, but it can be removed because `SecretKey` does the role of `Drop` of `DecrptedAccount`.

File: foundry/subcommand/convert_command.rs [hunk 15]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2019 Kodebox, Inc.
+// Copyright 2020 Kodebox, Inc.
```
Comment (sgkim126) [inline foundry/subcommand/convert_command.rs]: 2019-2020

File: key/src/ed25519/private.rs (hunk 13 above)
Comment (HoOngEe) [reply key/src/ed25519/private.rs]: Ok, I added a comment like "The inner type SecretKey clears its memory when it is dropped"

File: key/src/ed25519/private.rs (hunk 14 above)
Comment (HoOngEe) [reply key/src/ed25519/private.rs]: Ok that's great

File: key/src/ed25519/public.rs [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+    pub fn from_slice(slice: &[u8]) -> Option<Self> {
+        PublicKey::from_slice(slice).map(Self)
+    }
+
+    pub fn is_zero(&self) -> bool {
+        self.as_ref() == [0; PUBLICKEYBYTES]
+    }
+}
+
+impl Default for Public {
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply key/src/ed25519/public.rs]: Thank you for your scrutiny I removed it

File: key/src/ed25519/public.rs [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+    fn default() -> Self {
+        Public(PublicKey::from_slice(&[0; PUBLICKEYBYTES]).expect("Slice from an array with proper length never fails"))
+    }
+}
+
+impl From<u64> for Public {
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply key/src/ed25519/public.rs]: I tried to add `#[cfg(test)]` to the implementation. However, conditional complied test implementations were invisible in other crates. I hoped to do so but it seems no sharp solution is available.

File: key/src/lib.rs [hunk 18]
```diff
@@ -31,17 +31,16 @@ extern crate serde;
 #[macro_use]
 extern crate serde_derive;
 extern crate serde_json;
-extern crate sodiumoxide;
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply key/src/lib.rs]: Ok, I moved the change to the previous commit

File: foundry/subcommand/convert_command.rs (hunk 15 above)
Comment (HoOngEe) [reply foundry/subcommand/convert_command.rs]: Thanks, fixed.

File: core/src/account_provider.rs (hunk 4 above)
Comment (HoOngEe) [reply core/src/account_provider.rs]: I modified the commit message to be correct.

**Comment by sgkim126** [comment]
@HoOngEe Please check the e2e tests. It seems that some RLP encodings are wrong.

File: key/src/ed25519/public.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (sgkim126) [reply key/src/ed25519/public.rs]: Okay. It seems that we should leave it or replace the test code.
Let's leave it for now.

**Comment by HoOngEe** [comment]
https://github.com/CodeChain-io/foundry/pull/154#issuecomment-585041845
To pass e2e test, it requires ed25519 JavaScript implementation. What is the recommended plan?

**Comment by sgkim126** [comment]
@majecty @HoOngEe @kseo 
IMO, to skip tests will be dangerous since we have a plan to change many things.
How about finding the ed25519 implementation of JavaScript at this time?

**Comment by HoOngEe** [comment]
@sgkim126 We can utilize the https://github.com/jedisct1/libsodium.js/. After the sdk code moves into the foundry, @somniumism will introduce javascript `ed25519`.

**Comment by sgkim126** [comment]
It seems that the test still fails. @HoOngEe would you check it?

**Comment by HoOngEe** [comment]
@sgkim126 Sorry, it's not done yet. I'll finish it by today.

**Comment by HoOngEe** [comment]
I'm done until the tests triggered by `yarn start-short`. 
The remaining tests triggered by `yarn start-long` need a `derive` function to share secret keys in `p2pLayer`. The derive is previously just an ECDH. We can implement the counterpart with `tweetnacl` which is a libsodium wrapper. I think it's better to support the functionality in `foundry-primitives` for general purposes.
I also found some mishandled rejected expectations. I'll fix them in another pull request.

**Comment by HoOngEe** [comment]
I checked all the tests including dynamic validator tests pass in my local environment except the `snapshot` test in the `dyn-val-2`. It also fails in the master branch and it should be fixed in another pr.

**Comment by HoOngEe** [comment]
Unlike the first draft, I added two modules in the `key` crate.
1. Ed25519 keys and KeyPair for signatures
2. X25519 keys and KeyPair for key exchanges in p2p connection.

`key` crate exposes two key types with the prefixes `Ed25519` and `X25519`.
And each module imports such prefixed key types without prefix. For example, network related modules import` X25519KeyPair ` as `KeyPair` and other modules import  `Ed25519KeyPair` as `KeyPair`.

**Comment by HoOngEe** [comment]
However, it can be tested successfully with `foundry-primitives` version `0.2.0` after removing the concept of `Asset` in the test framwork. I'll make a pr for it.

File: test/src/stakeholder/actions/reportDoubleVote.ts [hunk 19]
```diff
@@ -46,16 +46,16 @@ export default class ReportDoubleVote {
...
+    // public criminal(): H512 {
+    //     const { message1 } = this;
+    //     const digest = blake256(RLP.encode(messageToEncodeObject(message1)[0]));
+    //     const signature = message1.signature.toString();
+    //     const pubkey = recoverSchnorr(digest, {
+    //         r: signature.slice(0, 64),
+    //         s: signature.slice(64)
+    //     });
+    //     return H512.ensure(pubkey);
+    // }
```
Comment (majecty) [inline test/src/stakeholder/actions/reportDoubleVote.ts]: If this code is not used, let's remove it.

Comment (HoOngEe) [reply test/src/stakeholder/actions/reportDoubleVote.ts]: Thank you, I removed it. I'll force-push after your review ends

Comment (majecty) [reply test/src/stakeholder/actions/reportDoubleVote.ts]: I reviewed it. Other code looks good to me.

Comment (HoOngEe) [reply test/src/stakeholder/actions/reportDoubleVote.ts]: Ok, I applied the change

File: test/src/e2e.dynval/1/doubleVote.test.ts [hunk 20]
```diff
@@ -73,7 +73,7 @@ describe("Double vote detection", function() {
             aliceNode.testFramework
         );
         expect(banned.map(b => b.toString())).to.include(
-            betty.platformAddress.toString()
+            betty.address.toString()
```
Comment (sgkim126) [inline test/src/e2e.dynval/1/doubleVote.test.ts]: Is the platform address just renamed to address, or was the type changed?

Comment (sgkim126) [reply test/src/e2e.dynval/1/doubleVote.test.ts]: If the type is changed, we don't use the platform address anymore?

Comment (HoOngEe) [reply test/src/e2e.dynval/1/doubleVote.test.ts]: It was renamed in the `foundry-primitives` after removing `AssetAddress`

**Comment by sgkim126** [comment]
@HoOngEe It seems that the snapshot test in dynamic-validator-2 still fails. Would you check it?

File: test/src/e2e.dynval/1/doubleVote.test.ts (hunk 20 above)
Comment (sgkim126) [reply test/src/e2e.dynval/1/doubleVote.test.ts]: If then, how do you call the 160 bits address that is the hash of the public key? In Foundry, it is address and the platform address is the combination of the network id, version, and address.

Comment (HoOngEe) [reply test/src/e2e.dynval/1/doubleVote.test.ts]: We were calling the `blake160(pub_key)` as `AccountId` in the sdk level. However, it is possible for one to be confused with the Foundry inner type `Address` and the higher level `Address`.

**Comment by HoOngEe** [comment]
@sgkim126 https://github.com/CodeChain-io/foundry/pull/154#issuecomment-595637945. I run the test in the master branch again, but still fails. I think we should fix the test in another pull request. Or am I missing something?

**Comment by sgkim126** [comment]
@HoOngEe The test successes on the master branch. The fragile test is one of the dyn-val-1.

**Comment by sgkim126** [comment]
And I re-ran the Travis several times at the weekend. It seems that the snapshot test fails completely. I suspect that the snapshot test uses ECDSA, but I didn't check it yet.

File: test/src/e2e.dynval/1/doubleVote.test.ts (hunk 20 above)
Comment (sgkim126) [reply test/src/e2e.dynval/1/doubleVote.test.ts]: Okay. Thank you for explanation.

**Comment by HoOngEe** [comment]
The failure caused by `0x` prefix of `blockHash` and `stateRoot`

**Comment by HoOngEe** [comment]
The root cause of the test failure is the difference between the types in `foundry-rpc-js` and the previous `codechain-sdk`. The hashes in the `block` of `foundry-rpc-js` have `string` types and they have `H256` types in the `sdk` (now they are moved into the test/src).
If you see [the code](https://github.com/CodeChain-io/foundry/blob/ef5d5664eea9125baf83b8d8f776914e0b281f2c/test/src/e2e.dynval/2/snapshot.test.ts#L82), each hash and stateRoot calls `toString()` method. Previously, `H256`'s `toString` method automatically remove `0x`prefixes but now the string instances with `0x` prefixes could not match with the actual snapshot path. Also, `self_nomination_enable` option should also be specified after the auto_self_nomination is merged into `foundry`. I'll add a fix commit

File: test/tendermint.dynval/snapshot-config.yml [hunk 21]
```diff
@@ -1,6 +1,7 @@
 [codechain]
 
 [mining]
+self_nomination_enable = false
```
Comment (sgkim126) [inline test/tendermint.dynval/snapshot-config.yml]: Why did you change it?

Comment (HoOngEe) [reply test/tendermint.dynval/snapshot-config.yml]: That's the quick fix for the error 
```
Error: "Error while parsing TOML: Error { inner: ErrorInner { kind: Custom, line: None, col: 0, message: \"missing field `self_nomination_enable`\", key: [\"mining\"] } }"
```
We also can designate default value for the field.
